### PR TITLE
Fixed issue #52

### PR DESCRIPTION
**Pull request by sepiatonal**
The chat is no longer spammed with messages from RealisticBiomes when you're harvesting a crop using the same type of crop in your hand.

### CODE REVIEW

**Comment by ttk2** [comment]
sweet, repeating my question from the other thread, is this good to merge?

**Comment by soerxpso** [comment]
Yes it is.

**Comment by ttk2** [comment]
cool, its on civtest

**Comment by ProgrammerDan** [comment]
@soerxpso looking at the commit a bit:

https://github.com/Civcraft/RealisticBiomes/pull/53/files#diff-45dadbac82aadd0a4b4d509bc3e57d7bR97

Are there any negatives to "dropping out" of the interaction event before performing the grow and persist method? E.g. with this change you can't force the crop to render at the accurate growth value more than once per second. For large fields or farms, this means you can't rapidly update them.

Other than placement, I think the code looks sound.

**Comment by soerxpso** [comment]
@ProgrammerDan It looks like you're correct. Doesn't it also work to just reload the chunk by relogging if you need the crops to re-render though? In any case you can change the placement if you think it needs changed. Just a simple ctrl+c ctrl+v.

File: src/com/untamedears/realisticbiomes/listener/PlayerListener.java [hunk 1]
```diff
@@ -45,10 +49,19 @@ public void onPlayerInteractEvent(PlayerInteractEvent event) {
 			return;
 		}
 		
-		Plant plant = null;
-		
 		Block block = event.getClickedBlock();
 		
+		if (((List<ItemStack>)block.getDrops()).get(0).getType() == event.getMaterial()) {
```
Comment (gipsy-king) [inline src/com/untamedears/realisticbiomes/listener/PlayerListener.java]: Surely this cast is unsafe and could be changed to something like `block.getDrops().iterator().next()`?

File: src/com/untamedears/realisticbiomes/listener/PlayerListener.java [hunk 2]
```diff
@@ -45,10 +49,19 @@ public void onPlayerInteractEvent(PlayerInteractEvent event) {
 			return;
 		}
 		
-		Plant plant = null;
-		
 		Block block = event.getClickedBlock();
 		
+		if (((List<ItemStack>)block.getDrops()).get(0).getType() == event.getMaterial()) {
+			suppression.put(event.getPlayer().getUniqueId(), System.currentTimeMillis());
```
Comment (gipsy-king) [inline src/com/untamedears/realisticbiomes/listener/PlayerListener.java]: @ProgrammerDan do you think it's okay that this lis never gets cleared? It should be filling up with players over the course of the day.

**Comment by ProgrammerDan** [comment]
What is our total unique player count? Unless over 10k in a single day i have no concerns.

@Soerxpso no, iirc the chunk load doesn't update all the crops, hence the hit command to force it to do so. 

As for who should make the edit, unfortunately this was prematurely merged before others could review it, hence yes someone else might have to fix it. Traditionally, however, on reviews the original poster of the pull takes ownership of corrections highlighted by reviewers. Asking them to fix it is in bad form (generally) unless the fix is unclear or outside the ability of the original poster.

**Comment by ttk2** [comment]
total unique? 40k range

**Comment by ProgrammerDan** [comment]
Ever had a day where all 40k visited? I'd guess you get a sample of about 1k each day?

**Comment by gipsy-king** [comment]
What about the NPE? I'm gonna revert @soerxpso and handle null and remove
unsafe cast.
El 14/8/2015 4:50 p. m., "Daniel Boston" <notifications@github.com>
escribió:

Ever had a day where all 40k visited? I'd guess you get a sample of about
1k each day?

—
Reply to this email directly or view it on GitHub
<https://github.com/Civcraft/RealisticBiomes/pull/53#issuecomment-131133540>
.

**Comment by ttk2** [comment]
getting logins per day is easy, uniques per day will require a little more
work and log fu.

**Comment by ProgrammerDan** [comment]
@gipsy-king VERY good catch on that NPE, totally missed it. Clearly wasn't thinking it through when reviewing.

Yeah, revert and repair is best option. Also, probably a good idea to give the concurrent hashmap an initial capacity of 1000 so there's very little backing-store-churn.

@ttk2 Was more a framing question; exact numbers not needed. 40k is still small enough that even if every single person logged in before restart, I'd not be overly worried about a UUID/Long map using up too much memory.

**Comment by ttk2** [comment]
@ProgrammerDan  once sharding is a thing we need to be more memory cautious, but for the time being we have 64gb of ram and MC won't use more than 10ish on its own because GC starts to take too long.

**Comment by ProgrammerDan** [comment]
@ttk2 Do we have some clear idea the hardware specs per shard? If we do, I could take that into account while working on mods, and pull together some impact analysis kind of stuff

**Comment by ttk2** [comment]
we can buy what we need, but ideally we would run three or four shards on a machine like the current one (3.7 ghz hyperthreaded quad core with 64gb ddr3 and 128bg ssd)

**Comment by ProgrammerDan** [comment]
Gives us a lot to work with, great. 

Might need to do some hardcore "overall" analysis esp. with the caching changes for citadel and other in-memory tracking.

**Comment by ttk2** [comment]
frankly if you need ssh you can have it.

File: src/com/untamedears/realisticbiomes/listener/PlayerListener.java (hunk 1 above)
Comment (sepiatonal) [reply src/com/untamedears/realisticbiomes/listener/PlayerListener.java:54]: According to the bukkit docs, it is a safe cast. And don't call me Shirley.

**Pull request merged**